### PR TITLE
gc_worker: make background gc compatible with multi-rocks db

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
Signed-off-by: SpadeA-Tang <u6748471@anu.edu.au>

### What is changed and how it works?

Issue Number: Ref #13406

What's Changed:

Make background gc compatible with Multi-Rocks DB configuration：
Change the parameter of `modify_on_kv_engine` from `Vec<Modify>` to `HashMap<u64, Vec<Modify>>` where u64 here represents the region id.  So that each region handles modifications separately.
Change the type of the return value of `kv_engine` from `Self::Local` to `Option<Self::Local>` to reflect the version of v1 and v2 where in v2 (multi-rocksdb version), we do not have this single global kv engine.

Add MultiRocksEngine for testing purpose.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/gc_worker/gc_worker.rs [hunk 1]
```diff
@@ -1483,6 +1519,118 @@ pub mod test_gc_worker {
...
+
+        fn kv_engine(&self) -> Self::Local {
+            panic!()
+        }
+
+        fn modify_on_kv_engine(
+            &self,
+            mut region_modifies: HashMap<u64, Vec<Modify>>,
+        ) -> kv::Result<()> {
+            for modifies in region_modifies.values_mut() {
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:1547]: Why loop twice?

File: src/server/gc_worker/gc_worker.rs [hunk 2]
```diff
@@ -367,18 +367,28 @@ where
...
+        if write_size > 0 {
             limiter.blocking_consume(write_size);
             engine.modify_on_kv_engine(modifies)?;
         }
         Ok(())
     }
 
     fn gc(&mut self, region: Region, safe_point: TimeStamp) -> Result<()> {
-        if !self.need_gc(region.get_start_key(), region.get_end_key(), safe_point) {
+        if !self.engine.multi_rocks()
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:389]: Why need to check if it's multi rocks?

File: src/server/gc_worker/gc_worker.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs:1547]: One for region id -> Vec<Modify>, and the other is for Modify in vector.

File: src/server/gc_worker/gc_worker.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs:389]: multi rocks has not this engine for this check.

File: src/server/gc_worker/gc_worker.rs (hunk 1 above)
Comment (BusyJay) [reply src/server/gc_worker/gc_worker.rs:1541]: Why not
```
for (region_id, modifies) in region_modifies {
    for modify in &mut modifies {
        xxx
    }
    write_modifies(factory.get_tablet(region_id), modifies);
}
```

**Review by BusyJay — COMMENTED** [review]
I think `need_gc` also needs to be changed to region based.

File: src/storage/txn/actions/tests.rs [hunk 3]
```diff
@@ -632,7 +680,33 @@ pub fn must_prewrite_delete<E: Engine>(
...
+        key,
+        pk,
+        ts,
+        TimeStamp::zero(),
+        SkipPessimisticCheck,
+        None,
+    );
+}
+
+pub fn must_prewrite_delete_v2<E: Engine>(
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply src/storage/txn/actions/tests.rs:694]: Good suggestion.

**Comment by tonyxuqqi** [comment]
Please add some explanation on what's the main change and  how this change could make it compatible with v2.  It's not very straightforward. The code itself is LGTM.

**Review by tonyxuqqi — APPROVED** [review]

File: components/tikv_kv/src/lib.rs [hunk 4]
```diff
@@ -267,10 +268,19 @@ pub trait Engine: Send + Clone + 'static {
     type Local: LocalEngine;
 
     /// Local storage engine.
-    fn kv_engine(&self) -> Self::Local;
+    ///
+    /// In the single rocksdb version, the return value is guaranteed to be
+    /// `Some` where as in the multi-rocksdb version, the `None` will be
+    /// returned.
+    fn kv_engine(&self) -> Option<Self::Local>;
```
Comment (tabokie) [inline components/tikv_kv/src/lib.rs:275]: How about `test_kv_engine` or `_kv_engine`, if this is only used in tests. (I'm not sure the idiomatic way to express this in Rust, cc @BusyJay)

File: src/server/raftkv.rs [hunk 5]
```diff
@@ -325,34 +326,43 @@ where
     type Snap = RegionSnapshot<E::Snapshot>;
     type Local = E;
 
-    fn kv_engine(&self) -> E {
-        self.engine.clone()
+    fn kv_engine(&self) -> Option<E> {
+        Some(self.engine.clone())
```
Comment (tabokie) [inline src/server/raftkv.rs:330]: Is there a followup PR? Because raftkv should be multi-rocksdb.

Comment (SpadeA-Tang) [reply src/server/raftkv.rs:330]: There will likely be a new v2 RaftKv struct which imlements `Engine` where no this kv_engine shoud be existed and it returns `None` instead.

Comment (tabokie) [reply src/server/raftkv.rs:330]: Is it possible to construct the `Engine` using tablet factory? Using something similar to `TestTabletFactory` can adapt to singleton case.

Comment (SpadeA-Tang) [reply src/server/raftkv.rs:330]: Use tablet factory is one choice, but after the PR https://github.com/tikv/tikv/pull/13322, we may no longer need tablet factory.

Comment (tabokie) [reply src/server/raftkv.rs:330]: Main thing is why introducing a RaftKvV2, isn't it used to hold a tablet factory?

Comment (SpadeA-Tang) [reply src/server/raftkv.rs:330]: I don't know exactly why introducing RaftKvV2. @BusyJay

Comment (BusyJay) [reply src/server/raftkv.rs:330]: No, v2 will not have a tablet factory.

File: components/tikv_kv/src/lib.rs (hunk 4 above)
Comment (BusyJay) [reply components/tikv_kv/src/lib.rs:275]: It will be used for v1.

File: src/server/gc_worker/gc_worker.rs [hunk 6]
```diff
@@ -367,10 +369,18 @@ where
...
     }
 
-    fn flush_txn(txn: MvccTxn, limiter: &Limiter, engine: &E) -> Result<()> {
-        let write_size = txn.write_size();
-        let modifies = txn.into_modifies();
-        if !modifies.is_empty() {
+    fn flush_txn(
+        limiter: &Limiter,
+        engine: &E,
+        txns: std::collections::HashMap<u64, MvccTxn>, // region id -> MvccTxn for this region
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:375]: Why use `std` version?

File: src/server/gc_worker/gc_worker.rs [hunk 7]
```diff
@@ -489,7 +501,10 @@ where
                     next_gc_key = keys_in_region.next();
                     gc_info = GcInfo::default();
                 } else {
-                    Self::flush_txn(txn, &self.limiter, &self.engine)?;
+                    txns.insert(region.id, txn);
+                    Self::flush_txn(&self.limiter, &self.engine, txns)?;
+                    txns = std::collections::HashMap::new();
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:506]: Why not `txns.clear()`?

File: src/server/gc_worker/gc_worker.rs [hunk 8]
```diff
@@ -2410,4 +2564,313 @@ mod tests {
...
+            .unwrap();
+
+        let mut region_id = 1;
+        let mut db = factory
+            .open_tablet(1, None, OpenOptions::default().set_cache_only(true))
+            .unwrap()
+            .as_inner()
+            .clone();
+        let mut cf = get_cf_handle(&db, CF_WRITE).unwrap();
+        for i in 0..30 {
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:2714]: You mean
```
for region_id in 1..=3 {
    for k in 10 * (region_id - 1)..10 * region_id {
    }
}
```
?

File: src/server/gc_worker/gc_worker.rs (hunk 7 above)
Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs:506]: It was moved.

File: src/server/gc_worker/gc_worker.rs (hunk 8 above)
Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs]: Yes

File: components/tikv_kv/src/lib.rs [hunk 9]
```diff
@@ -267,10 +268,19 @@ pub trait Engine: Send + Clone + 'static {
     type Local: LocalEngine;
 
     /// Local storage engine.
-    fn kv_engine(&self) -> Self::Local;
+    ///
+    /// In the single rocksdb version, the return value is guaranteed to be
+    /// `Some` where as in the multi-rocksdb version, the `None` will be
+    /// returned.
```
Comment (BusyJay) [inline components/tikv_kv/src/lib.rs:274]:
```suggestion
    /// If local engine can't be accessed directly, `None` is returned. Currently,
    /// only multi-rocksdb version will return `None`.
```

File: components/tikv_kv/src/lib.rs [hunk 10]
```diff
@@ -267,10 +268,19 @@ pub trait Engine: Send + Clone + 'static {
...
+    fn kv_engine(&self) -> Option<Self::Local>;
 
     /// Write modifications into internal local engine directly.
-    fn modify_on_kv_engine(&self, modifies: Vec<Modify>) -> Result<()>;
+    ///
+    /// region_modifies records each region's modifications. In the single
+    /// rocksdb version, they are flattened to be handled together whereas in
+    /// the multi-rocksdb version, the modifications of each region will be
+    /// handled separately.
```
Comment (BusyJay) [inline components/tikv_kv/src/lib.rs:282]:
```suggestion
    /// region_modifies records each region's modifications.
```

File: src/server/raftkv.rs [hunk 11]
```diff
@@ -325,34 +325,43 @@ where
...
-                    let bytes = keys::data_key(key1.as_encoded());
-                    *key1 = Key::from_encoded(bytes);
-                    let bytes = keys::data_end_key(key2.as_encoded());
-                    *key2 = Key::from_encoded(bytes);
+    fn modify_on_kv_engine(
+        &self,
+        mut region_modifies: HashMap<u64, Vec<Modify>>,
+    ) -> kv::Result<()> {
+        for modifies in region_modifies.values_mut() {
+            for modify in modifies.iter_mut() {
```
Comment (BusyJay) [inline src/server/raftkv.rs:337]:
```suggestion
            for modify in &mut modifies {
```

File: src/server/gc_worker/gc_worker.rs [hunk 12]
```diff
@@ -367,10 +369,18 @@ where
...
     }
 
-    fn flush_txn(txn: MvccTxn, limiter: &Limiter, engine: &E) -> Result<()> {
-        let write_size = txn.write_size();
-        let modifies = txn.into_modifies();
-        if !modifies.is_empty() {
+    fn flush_txn(
+        limiter: &Limiter,
+        engine: &E,
+        txns: HashMap<u64, MvccTxn>, // region id -> MvccTxn for this region
```
Comment (BusyJay) [inline src/server/gc_worker/gc_worker.rs:375]: How about using `&mut HashMap`? So it can be reused outside.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: src/server/gc_worker/gc_worker.rs (hunk 12 above)
Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs:375]: `into_modifies` of MvccTxn will be used in this method.

Comment (BusyJay) [reply src/server/gc_worker/gc_worker.rs:375]:
```
for (id, txn) in txns.drain() {
    txn.into_modifies()
}
```

Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs]: Good catch.

Comment (SpadeA-Tang) [reply src/server/gc_worker/gc_worker.rs]: Good suggestion.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**